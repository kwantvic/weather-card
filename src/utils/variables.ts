export const favorites: string = "arrFavCity";