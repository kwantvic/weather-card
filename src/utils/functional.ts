import {favorites} from "./variables";
import {FindLocationModel} from "../models/api/findLocation";
import {DetailedLocationModel} from "../models/api/detailedLocation";
import {AxiosError} from "axios";
import {ServerError} from "../models/redux/findCity";

function getLocalFav() {
    return localStorage.getItem(favorites);
}

function setLocalFav(arr: number[]) {
    return localStorage.setItem(favorites, JSON.stringify(arr));
}

function toTextualDescription(degree: number) {
    const sectors = ["Northerly", "North-Easterly", "Easterly", "South-Easterly", "Southerly", "South-Westerly", "Westerly", "North-Westerly"];
    degree += 22.5;
    (degree < 0) ? degree = 360 - Math.abs(degree) % 360 : degree = degree % 360;
    let which = parseInt(String(degree / 45));
    return sectors[which];
}

function mapLocationToCityState(obj: FindLocationModel) {
    return {
        id: obj.id,
        name: obj.name,
        country: obj.sys.country,
        temp: obj.main.temp,
        weathIcon: obj.weather[0].icon,
        weathDescr: obj.weather[0].description,
        lat: obj.coord.lat,
        lon: obj.coord.lon
    };
}

function mapDetailsToDetailedState(obj: DetailedLocationModel) {
    return {
        hourlyTemp: obj.hourly.slice(0, 12).map(({temp}: { temp: number }) => Math.round(temp)),
        details: {
            timezone: obj.timezone,
            uvi: obj.current.uvi,
            visibility: obj.current.visibility,
            windSpeed: obj.current.wind_speed,
            windDegree: obj.current.wind_deg,
            pressure: obj.current.pressure,
            tempFeels: obj.current.feels_like,
            sunrise: obj.current.sunrise,
            sunset: obj.current.sunset,
            humidity: obj.current.humidity
        }
    };
}

function getError(err: any) {
    let error: AxiosError<ServerError> = err;
    return error.response?.data ?? {message: "Error loading data"}
}

export {getLocalFav, setLocalFav, toTextualDescription, mapLocationToCityState, mapDetailsToDetailedState, getError};