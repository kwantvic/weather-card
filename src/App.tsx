import React from "react";
import {Routes, Route} from "react-router-dom";
import {useDispatch} from "react-redux";

import {Header} from "./components/Header";
import {ItemsFind} from "./components/ItemsCard/ItemsFind";
import {ActionAlerts} from "./components/UiComponents/ActionAlerts";
import {useErrorSelector} from "./redux/selectors";
import {ItemsFavorite} from "./components/ItemsCard/ItemsFavorite";
import {Detailed} from "./components/Detailed";
import {getLocalFav, setLocalFav} from "./utils/functional";
import {addLocalFav} from "./redux/favCitySlice";
import {NotFound} from "./components/NotFound";

export const App: React.FC = React.memo(() => {
    const dispatch = useDispatch();
    const errorSelector = useErrorSelector();

    React.useEffect(() => {
        if (getLocalFav()) {
            let arr: number[] = JSON.parse(getLocalFav()!);
            if (arr.length > 0)  {
                dispatch(addLocalFav(arr));
            }
        } else {
            setLocalFav([]);
        }
    }, [dispatch]);

    return (
        <div>
            <ActionAlerts severity={"error"} errorDescription={errorSelector}/>
            <Header/>
            <Routes>
                <Route path="/" element={<ItemsFavorite/>}/>
                <Route path="/find/:location" element={<ItemsFind/>}/>
                <Route path="/detailed/:id" element={<Detailed/>}/>
                <Route path="*" element={<NotFound/>}/>
            </Routes>
        </div>
    );
})