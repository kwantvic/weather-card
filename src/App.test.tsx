import React from "react";
import {render, screen} from "@testing-library/react";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";

import {App} from "./App";
import store from "./redux";
import {getLocalFav} from "./utils/functional";

jest.mock("./utils/functional", () => ({
	...(jest.requireActual("./utils/functional")),
	getLocalFav: jest.fn()
}));

describe("APP", () => {

	it("renders APP", () => {
		render(<MemoryRouter><Provider store={store}><App/></Provider></MemoryRouter>);
		expect(screen.getByRole("textbox")).toBeInTheDocument();
	});

	it("snapshot APP without favItems", () => {
		expect(render(<MemoryRouter><Provider store={store}><App/></Provider></MemoryRouter>)).toMatchSnapshot();
	});

	it("renders APP without local favorites cities", () => {
		(getLocalFav as jest.Mock).mockReturnValue("[]");
		expect(getLocalFav()).toEqual("[]");
		render(<MemoryRouter><Provider store={store}><App/></Provider></MemoryRouter>);
		expect(screen.queryByTestId("city-preload")).not.toBeInTheDocument();
	});

	it("renders APP with local favorites cities", () => {
		(getLocalFav as jest.Mock).mockReturnValue("[1111, 2222]");
		expect(getLocalFav()).toEqual("[1111, 2222]");
		render(<MemoryRouter><Provider store={store}><App/></Provider></MemoryRouter>);
		expect(screen.getAllByTestId("city-preload")[0]).toBeInTheDocument();
	});
})