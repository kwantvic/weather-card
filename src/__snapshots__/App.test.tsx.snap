// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP snapshot APP without favItems 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="wrapper false"
          data-testid="error-alert"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="ErrorOutlineIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              <div
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root error css-yslrua-MuiTypography-root-MuiAlertTitle-root"
                style="text-transform: capitalize;"
              >
                error
              </div>
              <span>
                 ðŸ¥º
              </span>
            </div>
          </div>
        </div>
        <div
          class="wrapper"
        >
          <div
            class="header"
          >
            <span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
                data-testid="SettingsSystemDaydreamIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 16h6.5c1.38 0 2.5-1.12 2.5-2.5S16.88 11 15.5 11h-.05c-.24-1.69-1.69-3-3.45-3-1.4 0-2.6.83-3.16 2.02h-.16C7.17 10.18 6 11.45 6 13c0 1.66 1.34 3 3 3zM21 3H3c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16.01H3V4.99h18v14.02z"
                />
              </svg>
               weather-card
            </span>
          </div>
          <div
            class="search"
          >
            <div
              class="form"
            >
              <input
                class="input"
                placeholder="Weather in your city"
                type="text"
                value=""
              />
              <button
                class="btn btnColor"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium iconSearch css-i4bv87-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
                Search
              </button>
            </div>
          </div>
        </div>
        <div
          class="wrapper"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-12mvehg-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="InfoOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              Here you can add a card with the current weather of your city ðŸ˜‰
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="wrapper false"
        data-testid="error-alert"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="ErrorOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message css-acap47-MuiAlert-message"
          >
            <div
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root error css-yslrua-MuiTypography-root-MuiAlertTitle-root"
              style="text-transform: capitalize;"
            >
              error
            </div>
            <span>
               ðŸ¥º
            </span>
          </div>
        </div>
      </div>
      <div
        class="wrapper"
      >
        <div
          class="header"
        >
          <span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
              data-testid="SettingsSystemDaydreamIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16h6.5c1.38 0 2.5-1.12 2.5-2.5S16.88 11 15.5 11h-.05c-.24-1.69-1.69-3-3.45-3-1.4 0-2.6.83-3.16 2.02h-.16C7.17 10.18 6 11.45 6 13c0 1.66 1.34 3 3 3zM21 3H3c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16.01H3V4.99h18v14.02z"
              />
            </svg>
             weather-card
          </span>
        </div>
        <div
          class="search"
        >
          <div
            class="form"
          >
            <input
              class="input"
              placeholder="Weather in your city"
              type="text"
              value=""
            />
            <button
              class="btn btnColor"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium iconSearch css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
              Search
            </button>
          </div>
        </div>
      </div>
      <div
        class="wrapper"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-12mvehg-MuiPaper-root-MuiAlert-root"
          role="alert"
        >
          <div
            class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="InfoOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message css-acap47-MuiAlert-message"
          >
            Here you can add a card with the current weather of your city ðŸ˜‰
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
